feat(embedded): allow resetting filters

ifrost opened the pull request:
Allows removing/changing/resetting (previously manged) filters in Embedded mode. This will also be used to switch over to Logs tab when a data link is clicked in Asserts:

https://github.com/user-attachments/assets/d4e2df49-4d48-470f-8508-a054070921a1

All PRs related to this change:

- [ ] https://github.com/grafana/logs-drilldown/pull/1549
- [ ] https://github.com/grafana/grafana/pull/111304
- [ ] https://github.com/grafana/asserts-app-plugin/pull/2150

gtk-grafana commented:
@ifrost I made the change to remove the readonly filters in https://github.com/grafana/logs-drilldown/pull/1547 as well, do you have a preference if we address that change in this PR or not?

ifrost commented:
> @ifrost I made the change to remove the readonly filters in #1547 as well, do you have a preference if we address that change in this PR or not?

Oh, nice! Feel free to merge the other one first if it can speed things up for you. I've additionally:
- removed some (hopefully) dead code, 
- added a reset button (used by https://github.com/grafana/asserts-app-plugin/pull/2150),
- added handling for the case where user removes all filters (there's no service selection in embedded view and it would throw an error). 
- added ability to pass custom namespace for https://github.com/grafana/asserts-app-plugin/pull/2150 (though can hardcode `ld` as well)

ifrost commented on src/Components/IndexScene/IndexScene.tsx:
```diff
@@ -300,6 +305,13 @@ export class IndexScene extends SceneObjectBase<IndexSceneState> {
+  public currentFiltersMatchReference() {
+    return (
+      !this.state.referenceLabels ||
+      areLabelFiltersEqual(this.state.referenceLabels || [], getLabelsVariable(this).state.filters)
```
Nice! That should do the trick and be simpler 👍

ifrost replied:
Updated in ee989448770718e98302aeeca16191b7e207aa2b

gtk-grafana commented:
@ifrost I'm seeing that the reset filters button always removes all filters and pushes to the `No labels selected` state when testing locally. Does this require changes to the props passed in from embedding application? If so let's update the [testing route](https://github.com/grafana/logs-drilldown/blob/main/src/Components/Pages.tsx#L76-L112) so this PR can be tested without external deps

gtk-grafana commented on src/services/variableHelpers.ts:
```diff
@@ -1,38 +1,24 @@
-        variablesToClear.push(variable);
-      }
-    } else if (variable instanceof AdHocFiltersVariable && variable.state.filters.length) {
+    if (variable instanceof AdHocFiltersVariable && variable.state.filters.length) {
```
@ifrost I think this change is what's causing the initial labels to get removed (and most likely also a bug in my other branch removing the read only filters), we were relying on the readonly variable to let us know which filters should get left alone when clearing.

ifrost commented:
Now it should not require external deps - with Asserts from main it will just not show the reset button at all. If used with https://github.com/grafana/asserts-app-plugin/pull/2150 the reference labels will be passed from Asserts and the reset button will show up.

gtk-grafana commented on src/Components/EmbeddedLogsExploration/EmbeddedLogs.tsx:
```diff
@@ -34,13 +35,20 @@ export function buildLogsExplorationFromState({
   initRuntimeDs();
 
   const { labelFilters, lineFilters } = getMatcherFromQuery(query);
+  const referenceFilters = getMatcherFromQuery(referenceQuery || '');
```
```suggestion
  const referenceFilters = getMatcherFromQuery(referenceQuery ?? '');
```

gtk-grafana commented on src/Components/EmbeddedLogsExploration/EmbeddedLogs.tsx:
```diff
@@ -74,7 +83,7 @@ export default function EmbeddedLogsExploration(props: EmbeddedLogsExplorationPr
       updateUrlOnInit={false}
       createBrowserHistorySteps={true}
-      namespace={VARIABLE_NAMESPACE}
+      namespace={props.namespace || VARIABLE_NAMESPACE}
```
```suggestion
      namespace={props.namespace ?? VARIABLE_NAMESPACE}
```

gtk-grafana reviewed (commented):
Everything is looking good for me locally, a few super minor nits (we use null coalesce here) in the code, but this looks good to me. 

@matyax will you have a few minutes today to review? It would be good to get this merged ASAP and get a deploy out early next week

ifrost commented:
> @ifrost I added the referenceQuery to the testing route props in [33cd4b0](https://github.com/grafana/logs-drilldown/commit/33cd4b0ac00ab92c088d9d289acc4465bbbb6128). The testing route can be hit at `/a/grafana-lokiexplore-app/embed`
> 
> By external deps I mean PRs in external repos, the testing route is supposed to save the rest of the squad from needing to spin up Logs Drilldown in an external app, and to provide us something to run e2e tests against that should provide the same behavior as within an embedded app (e.g. Asserts).

Oh that's neat! Didn't know about it. It'd be useful be useful in other apps 👍

matyax commented on src/Components/EmbeddedLogsExploration/EmbeddedLinkScene.tsx:
```diff
@@ -41,6 +41,13 @@ export class EmbeddedLinkScene extends SceneObjectBase {
+    // The link to breakdown cannot be created without labels
+    // Alternatively we could build a link to ServiceSelectionScene
+    if (labelsVarFilters.length === 0) {
+      return null;
```
Worth adding a warning logger to not silently fail.

matyax commented on src/Components/IndexScene/VariableLayoutScene.tsx:
```diff
@@ -99,6 +99,16 @@ export class VariableLayoutScene extends SceneObjectBase<VariableLayoutSceneStat
                     ) : null;
                   })}
                 </div>
+                {!currentFiltersMatchReference && (
```
Just do you're aware, this can be potentially displayed if the reference filters are not defined.


```typescript
public currentFiltersMatchReference() {
    const referenceLabelsDefined = this.state.referenceLabels && this.state.referenceLabels.length > 0;
    return (
      !referenceLabelsDefined || ...
```

gtk-grafana replied:
Yeah, although currently if not embedded it will redirect to the service selection scene, but embedded without reference would display this button (which does nothing)

matyax reviewed (commented):
Just a couple of minor comments, but LGTM!

gtk-grafana reviewed (commented):
Can we add a copy of the reset button to this warning?
<img width="965" height="94" alt="Image" src="https://github.com/user-attachments/assets/15ae9dfc-d75f-46d7-ab82-8790782dc9fb" />

something like:

<img width="1273" height="115" alt="Image" src="https://github.com/user-attachments/assets/c48ee9ba-ea3b-4003-89fb-75d8d74673db" />

ifrost commented:
> @ifrost pretty big regression in the [last commit](https://github.com/grafana/logs-drilldown/pull/1549/commits/b9a06b47dcd2083915f583e988b0a0c2398acd86)

Yep, keeping "filters validity" in state was a mistake. The state gets stale when moving back and forth between views (removing labels, or using browser history) and would require cleaning up which may feel clunky. In 7a2c847dd9d33600d7638c9fe0acd2981fd93952 I have switched from storing the state to deriving it in runtime, and split the the handling logic from the subscription handler to separate methods.

It's been a few back and forth with this PR. If it's gonna be easier I can split it into smaller PRs and start with the refactoring only 🤔

gtk-grafana commented:
Sounds good, I ran into this error while testing: https://github.com/grafana/logs-drilldown/pull/1552, but that exists on main and is not related to your changes.

Looks good to me, everything is working locally, although I agree it's a bit messy, but refactoring in https://github.com/grafana/logs-drilldown/pull/1551 should work fine.

@matyax Soft approval from me, feel free to approve and merge in the morning if you have some time, otherwise I can give final approval tomorrow afternoon.

gtk-grafana commented:
oh @ifrost please sign the CLA again, I don't know why it expires

matyax commented on src/Components/IndexScene/ResetFiltersButton.tsx:
```diff
@@ -0,0 +1,26 @@
+        icon="repeat"
+        variant="secondary"
+        onClick={() => indexScene.resetToReferenceQuery()}
+        tooltip="Reset filters to initial values."
```
There's just one thing that I noticed while testing this, is that we should be more explicit with the outcome. It will reset the label filters, but it's not being mentioned.

```suggestion
        tooltip="Reset label filters to initial values."
```

<img width="652" height="199" alt="Image" src="https://github.com/user-attachments/assets/cea1452c-3b39-4807-a169-f2df06f4966b" />

Labels reset, but other filters are kept.

<img width="622" height="196" alt="Image" src="https://github.com/user-attachments/assets/4fd6dc62-b921-4a87-a3e5-ad94f5becf42" />

matyax replied:
If you don't mind, I'll apply the suggestion so we can move forward with this.

matyax commented on src/Components/EmbeddedLogsExploration/EmbeddedLinkScene.tsx:
```diff
@@ -26,7 +26,7 @@ export class EmbeddedLinkScene extends SceneObjectBase {
 
     // Rerender this scene whenever any dependent variables are updated
```
Outdated comment.

ifrost replied:
Changed in 07300e948fbe1d8275700fa58ad7c8650cbeee0a

matyax commented on src/Components/ServiceScene/ServiceScene.tsx:
```diff
@@ -761,6 +810,27 @@ export class ServiceScene extends SceneObjectBase<ServiceSceneState> {
+
+    if (!status.isValid) {
+      return (
+        <Alert title="Invalid labels selected" severity="info">
```
```suggestion
        <Alert title={status.reason === LabelFiltersInvalidReason.Empty ? "No labels selected" : "Invalid labels selected"} severity="info">
```

ifrost commented:
Adressed [comment](https://github.com/grafana/logs-drilldown/pull/1551#discussion_r2381749732) in 333a24809e16bb689c1b318faa868b590033fe05

matyax reviewed (approved):
👌
